S/FTP sink materializes a successful IOResult if connection fails during upload

Hi there,

We are having an issue with the Sftp sink. It is reporting a successful upload in the event of an IOException happening during the upload its self.

Heres the test app Im using 

```scala
import java.net.InetAddress

import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import akka.stream.alpakka.ftp.FtpCredentials.NonAnonFtpCredentials
import akka.stream.alpakka.ftp.SftpSettings
import akka.stream.alpakka.ftp.scaladsl.Sftp
import akka.stream.scaladsl.{Keep, Source}
import akka.util.ByteString

object TestFtpSink extends App {
  private val host = InetAddress.getByName(args(0))
  private val path = args(1)
  private val username = args(2)
  private val password = args(3)

  implicit val actorSystem = ActorSystem("test")
  implicit val materializer = ActorMaterializer()
  implicit val ec = actorSystem.dispatcher

  val settings = SftpSettings(host = host,
                              credentials = NonAnonFtpCredentials(username, password),
                              strictHostKeyChecking = false)

  private val future = Source.repeat(ByteString(0x00)).toMat(Sftp.toPath(path, settings))(Keep.right).run()

  future.onComplete { result ⇒
    println(result)
    actorSystem.terminate()
  }
}
```
```scala
lazy val root = (project in file("."))
  .settings(
    name := "graph-stage-test",
    scalaVersion := "2.12.2",
    libraryDependencies += "com.lightbend.akka" %% "akka-stream-alpakka-ftp" % "0.10"
  )
```
The steps Im using to reproduce are:
* I run the above app against the sftp sever
* Wait until the file has been created remotely
* Kill the network on the local machine 
* wait a few seconds

The output I see is as follows:

```
[ERROR] [07/05/2017 11:15:05.281] [test-akka.stream.default-blocking-io-dispatcher-11] [akka://test/user/StreamSupervisor-0/flow-0-1-sFtpIOSink] Error in stage [akka.stream.alpakka.ftp.impl.FtpSourceFactory$$anon$3@16e96c32]: Broken pipe (Write failed)
java.net.SocketException: Broken pipe (Write failed)
	at java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:109)
	at java.net.SocketOutputStream.write(SocketOutputStream.java:153)
	at com.jcraft.jsch.IO.put(IO.java:60)
	at com.jcraft.jsch.Session._write(Session.java:1379)
	at com.jcraft.jsch.Session.write(Session.java:1343)
	at com.jcraft.jsch.ChannelSftp.sendWRITE(ChannelSftp.java:2646)
	at com.jcraft.jsch.ChannelSftp.access$100(ChannelSftp.java:36)
	at com.jcraft.jsch.ChannelSftp$1.write(ChannelSftp.java:806)
	at com.jcraft.jsch.ChannelSftp$1.write(ChannelSftp.java:789)
	at akka.stream.alpakka.ftp.impl.FtpIOSinkStage$$anon$2.$anonfun$write$1(FtpIOGraphStage.scala:182)
	at akka.stream.alpakka.ftp.impl.FtpIOSinkStage$$anon$2.$anonfun$write$1$adapted(FtpIOGraphStage.scala:181)
	at scala.Option.foreach(Option.scala:257)
	at akka.stream.alpakka.ftp.impl.FtpIOSinkStage$$anon$2.akka$stream$alpakka$ftp$impl$FtpIOSinkStage$$anon$$write(FtpIOGraphStage.scala:181)
	at akka.stream.alpakka.ftp.impl.FtpIOSinkStage$$anon$2$$anon$4.onPush(FtpIOGraphStage.scala:141)
	at akka.stream.impl.fusing.GraphInterpreter.processPush(GraphInterpreter.scala:747)
	at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:710)
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:616)
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:471)
	at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:410)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:603)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:618)
	at akka.actor.Actor.aroundReceive(Actor.scala:502)
	at akka.actor.Actor.aroundReceive$(Actor.scala:500)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:529)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

[ERROR] [07/05/2017 11:15:05.287] [test-akka.stream.default-blocking-io-dispatcher-11] [akka://test/user/StreamSupervisor-0/flow-0-1-sFtpIOSink] Error during postStop in [akka.stream.alpakka.ftp.impl.FtpSourceFactory$$anon$3@16e96c32]: inputstream is closed
java.io.IOException: inputstream is closed
	at com.jcraft.jsch.ChannelSftp.fill(ChannelSftp.java:2911)
	at com.jcraft.jsch.ChannelSftp.header(ChannelSftp.java:2935)
	at com.jcraft.jsch.ChannelSftp.checkStatus(ChannelSftp.java:2473)
	at com.jcraft.jsch.ChannelSftp.access$300(ChannelSftp.java:36)
	at com.jcraft.jsch.ChannelSftp$1.flush(ChannelSftp.java:851)
	at com.jcraft.jsch.ChannelSftp$1.close(ChannelSftp.java:867)
	at akka.stream.alpakka.ftp.impl.FtpIOSinkStage$$anon$2.$anonfun$postStop$2(FtpIOGraphStage.scala:163)
	at akka.stream.alpakka.ftp.impl.FtpIOSinkStage$$anon$2.$anonfun$postStop$2$adapted(FtpIOGraphStage.scala:163)
	at scala.Option.foreach(Option.scala:257)
	at akka.stream.alpakka.ftp.impl.FtpIOSinkStage$$anon$2.postStop(FtpIOGraphStage.scala:163)
	at akka.stream.impl.fusing.GraphInterpreter.finalizeStage(GraphInterpreter.scala:801)
	at akka.stream.impl.fusing.GraphInterpreter.afterStageHasRun(GraphInterpreter.scala:780)
	at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:620)
	at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:471)
	at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:410)
	at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:603)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:618)
	at akka.actor.Actor.aroundReceive(Actor.scala:502)
	at akka.actor.Actor.aroundReceive$(Actor.scala:500)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:529)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

Success(IOResult(71198,Success(Done)))

Process finished with exit code 0
```

This seems incorrect, I would expect the IOResult materialized by the stream to contain a Failure wrapping the original SocketException

--- a/ftp/src/main/scala/akka/stream/alpakka/ftp/impl/FtpIOGraphStage.scala
+++ b/ftp/src/main/scala/akka/stream/alpakka/ftp/impl/FtpIOGraphStage.scala
@@ -138,19 +138,21 @@ private[ftp] trait FtpIOSinkStage[FtpClient, S <: RemoteFileSettings]
       setHandler(
         in,
         new InHandler {
-          override def onPush(): Unit =
+          override def onPush(): Unit = {
             try {
               write(grab(in))
-              pull(in)
             } catch {
               case NonFatal(e) ⇒
                 matFailure(e)
                 try osOpt.foreach(_.close())
-                catch { case NonFatal(_) ⇒ }
+                catch {
+                  case NonFatal(_) ⇒
+                }
                 osOpt = None
                 throw e
             }
-
+            pull(in)
+          }
           override def onUpstreamFinish(): Unit =
             try {
               osOpt.foreach(_.close())

--- a/ftp/src/main/scala/akka/stream/alpakka/ftp/impl/FtpIOGraphStage.scala
+++ b/ftp/src/main/scala/akka/stream/alpakka/ftp/impl/FtpIOGraphStage.scala
@@ -11,6 +11,7 @@ import akka.util.ByteString
 import akka.util.ByteString.ByteString1C
 import scala.concurrent.{Future, Promise}
 import java.io.{InputStream, OutputStream}
+import scala.util.control.NonFatal
 
 private[ftp] trait FtpIOGraphStage[FtpClient, S <: RemoteFileSettings, Sh <: Shape]
     extends GraphStageWithMaterializedValue[Sh, Future[IOResult]] {
@@ -137,10 +138,18 @@ private[ftp] trait FtpIOSinkStage[FtpClient, S <: RemoteFileSettings]
       setHandler(
         in,
         new InHandler {
-          override def onPush(): Unit = {
-            write(grab(in))
-            pull(in)
-          }
+          override def onPush(): Unit =
+            try {
+              write(grab(in))
+              pull(in)
+            } catch {
+              case NonFatal(e) ⇒
+                matFailure(e)
+                try osOpt.foreach(_.close())
+                catch { case NonFatal(_) ⇒ }
+                osOpt = None
+                throw e
+            }
 
           override def onUpstreamFinish(): Unit =
             try {

--- a/ftp/src/test/scala/akka/stream/alpakka/ftp/CommonFtpStageSpec.scala
+++ b/ftp/src/test/scala/akka/stream/alpakka/ftp/CommonFtpStageSpec.scala
@@ -15,6 +15,7 @@ import scala.util.Random
 import java.nio.file.attribute.PosixFilePermission
 import java.nio.file.{Files, Paths}
 import java.net.InetAddress
+import org.scalatest.concurrent.Eventually
 
 final class FtpStageSpec extends BaseFtpSpec with CommonFtpStageSpec
 final class SftpStageSpec extends BaseSftpSpec with CommonFtpStageSpec
@@ -65,7 +66,7 @@ final class StrictHostCheckingSftpSourceSpec extends BaseSftpSpec with CommonFtp
   )
 }
 
-trait CommonFtpStageSpec extends BaseSpec {
+trait CommonFtpStageSpec extends BaseSpec with Eventually {
 
   implicit val system = getSystem
   implicit val mat = getMaterializer
@@ -242,6 +243,22 @@ trait CommonFtpStageSpec extends BaseSpec {
       val result = brokenSource.runWith(storeToPath(s"/$fileName", append = false)).futureValue
 
       result.status.failed.get shouldBe a[ArithmeticException]
+    }
+
+    "fail and report the exception in the result status if connection fails" in {
+      val fileName = "sample_io"
+      val infiniteSource = Source.repeat(ByteString(0x00))
+
+      val future = infiniteSource.runWith(storeToPath(s"/$fileName", append = false))
+      eventually {
+        noException should be thrownBy getFtpFileContents(FtpBaseSupport.FTP_ROOT_DIR, fileName)
+        getFtpFileContents(FtpBaseSupport.FTP_ROOT_DIR, fileName).length shouldBe >(0)
+      }
+      stopServer()
+      val result = future.futureValue
+      startServer()
+
+      result.status.failed.get shouldBe a[Exception]
     }
   }
 }

--- a/ftp/src/test/scala/akka/stream/alpakka/ftp/CommonFtpStageSpec.scala
+++ b/ftp/src/test/scala/akka/stream/alpakka/ftp/CommonFtpStageSpec.scala
@@ -246,14 +246,17 @@ trait CommonFtpStageSpec extends BaseSpec with Eventually {
     }
 
     "fail and report the exception in the result status if connection fails" in {
+      def waitForUploadToStart(fileName: String) =
+        eventually {
+          noException should be thrownBy getFtpFileContents(FtpBaseSupport.FTP_ROOT_DIR, fileName)
+          getFtpFileContents(FtpBaseSupport.FTP_ROOT_DIR, fileName).length shouldBe >(0)
+        }
+
       val fileName = "sample_io"
       val infiniteSource = Source.repeat(ByteString(0x00))
 
       val future = infiniteSource.runWith(storeToPath(s"/$fileName", append = false))
-      eventually {
-        noException should be thrownBy getFtpFileContents(FtpBaseSupport.FTP_ROOT_DIR, fileName)
-        getFtpFileContents(FtpBaseSupport.FTP_ROOT_DIR, fileName).length shouldBe >(0)
-      }
+      waitForUploadToStart(fileName)
       stopServer()
       val result = future.futureValue
       startServer()

